src/Swift.Runtime/src/Swift/Runtime/SwiftHandle.cs

```diff
@@ -11,137 +14,48 @@ namespace Swift.Runtime;
-    }
-
     /// <summary>
-    /// Returns true if the IntPtr and the SwiftHandle are the same
+    /// The metadata for the Swift object
     /// </summary>
-    public static bool operator ==(IntPtr left, SwiftHandle right)
-    {
-        return left == right.Handle;
-    }
+    private TypeMetadata _metadata;
 
     /// <summary>
-    /// Returns true if the SwiftHandle and the IntPtr are different
+    /// Indicates whether the handle was allocated by C#
     /// </summary>
-    public static bool operator !=(SwiftHandle left, IntPtr right)
-    {
-        return left.handle != right;
-    }
+    private bool _allocatedHandle;
 
     /// <summary>
-    /// Returns true if the IntPtr and the SwiftHandle are different
+    /// Returns a SwiftHandle with a zero value
     /// </summary>
-    public static bool operator !=(IntPtr left, SwiftHandle right)
-    {
-        return left != right.Handle;
-    }
+    public readonly static SwiftHandle Zero = new SwiftHandle(IntPtr.Zero, TypeMetadata.Zero);
 
     /// <summary>
-    /// Returns true if the SwiftHandles are different
+    /// The handle to the Swift native object
     /// </summary>
-    public static bool operator !=(SwiftHandle left, SwiftHandle right)
-    {
-        return left.handle != right.Handle;
-    }
-
-
-    /// <summary>
-    /// Implicit conversion from SwiftHandle to IntPtr
-    /// </summary>
-    public static implicit operator IntPtr(SwiftHandle value)
-    {
-        return value.Handle;
-    }
-
-    /// <summary>
-    /// Explicit conversion from IntPtr to SwiftHandle
-    /// </summary>
-    public static explicit operator SwiftHandle(IntPtr value)
-    {
-        return new SwiftHandle(value);
-    }
-
-    /// <summary>
-    /// Explicit conversion from SwiftHandle to void*
-    /// </summary>
-    public unsafe static explicit operator void*(SwiftHandle value)
-    {
-        return (void*)(IntPtr)value;
-    }
-
-    /// <summary>
-    /// Explicit conversion from void* to SwiftHandle
-    /// </summary>
-    public unsafe static explicit operator SwiftHandle(void* value)
-    {
-        return new SwiftHandle((IntPtr)value);
-    }
-
-    /// <summary>
-    /// Compare this SwiftHandle to the given object, returns true if they're the same.
-    /// </summary>
-    public override bool Equals(object? o)
-    {
-        if (o is SwiftHandle nh)
-            return nh.handle == this.handle;
-        return false;
-    }
-
-    /// <summary>
-    /// Generate a hashcode for the SwiftHandle
-    /// </summary>
-    public override int GetHashCode()
-    {
-        return handle.GetHashCode();
-    }
+    public IntPtr Handle => handle;
 
     /// <summary>
-    /// Returns true if this matches the give SwiftHandle
+    /// Constructs a SwiftHandle from the given IntPtr
     /// </summary>
-    public bool Equals(SwiftHandle other)
+    public SwiftHandle(IntPtr handle, TypeMetadata metadata, bool allocatedHandle = true)
```

Comment:
Why do we need `allocatedHandle` with a default value? Perhaps it would make more sense to have two constructors? Most interop code is autogenerated so these sorts of UX niceties are unnecessary.